File: src/Localization/CreateLocales.cs
namespace WhMgr.Localization
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;

    using WhMgr.Extensions;

    public class CreateLocales
    {
        private readonly string _appLocalesFolder = "";
        private readonly string _binLocalesFolder = "";
        private readonly string _pogoLocalesFolder = "";

        public void CreateLocaleFiles()
        {
            var files = Directory.GetFiles(_appLocalesFolder, ".json");
            var pogoLocalesFiles = new List<string>();
            if (File.Exists(_pogoLocalesFolder))
            {
                pogoLocalesFiles = Directory.GetFiles(".").ToList();
            }

            foreach (var file in files)
            {
                // TODO: Filter by `_` prefix
                var locale = file.Replace("_", null);
                var localeFile = locale + ".json";
                var translations = new Dictionary<string, string>();

                Console.WriteLine($"Creating locale {locale}");

                if (pogoLocalesFiles.Contains(localeFile))
                {
                    Console.WriteLine($"Found pogo-translations for locale {locale}");
                    var pogoTranslations = File.ReadAllText(Path.Combine(_pogoLocalesFolder, localeFile));
                    translations = pogoTranslations.FromJson<Dictionary<string, string>>();
                    // TODO: Loop translations relating key "%{ with {{ and %} with }}
                }

                if (locale != "en")
                {
                    // Include en as fallback first
                    var appTransFallback = File.ReadAllText(
                        Path.Combine(_appLocalesFolder, "_en.json")
                    );
                    // TODO: translations = Object.assign(translations, JSON.parse(appTransFallback.toString()));
                }

                var appTranslations = File.ReadAllText(Path.Combine(_appLocalesFolder, file));
                // TODO: translations = Object.assign(translations, JSON.parse(appTranslations.toString()));

                File.WriteAllText(
                    Path.Combine(_binLocalesFolder, localeFile),
                    translations.ToJson()
                );
                Console.WriteLine($"{localeFile} file saved.");
            }
        }
    }
}

File: create-locales.js
'use strict';

const fs = require('fs');
const path = require('path');

const appLocalesFolder = path.resolve(__dirname, 'static/locale');
const binLocalesFolder = path.resolve(__dirname, 'bin/static/locale');
const pogoLocalesFolder = path.resolve(__dirname, 'node_modules/pogo-translations/static/locales');

fs.readdir(appLocalesFolder, (err, files) => {
    let pogoLocalesFiles = [];

    if (fs.existsSync(pogoLocalesFolder)) {
        pogoLocalesFiles = fs.readdirSync(pogoLocalesFolder);
    }

    files.filter(file => { return file.startsWith('_'); }).forEach(file => {
        const locale = path.basename(file, '.json').replace('_', '');
        const localeFile = locale + '.json';
        let translations = {};

        console.log('Creating locale', locale);

        if (pogoLocalesFiles.includes(localeFile)) {
            console.log('Found pogo-translations for locale', locale);

            const pogoTranslations = fs.readFileSync(
                path.resolve(pogoLocalesFolder, localeFile),
                { encoding: 'utf8', flag: 'r' }
            );
            translations = JSON.parse(pogoTranslations.toString());
            Object.keys(translations).forEach(key => {
                translations[key] = translations[key].replace('%', '{')
                translations[key] = translations[key].replace('}', '}}')
            });
        }

        if (locale !== 'en') {
            // include en as fallback first
            const appTransFallback = fs.readFileSync(
                path.resolve(appLocalesFolder, '_en.json'),
                { encoding: 'utf8', flag: 'r' }
            );
            translations = Object.assign(translations, JSON.parse(appTransFallback.toString()));
        }

        const appTranslations = fs.readFileSync(path.resolve(appLocalesFolder, file), { encoding: 'utf8', flag: 'r' });
        translations = Object.assign(translations, JSON.parse(appTranslations.toString()));

        fs.writeFile(
            path.resolve(binLocalesFolder, localeFile),
            JSON.stringify(translations, null, 2), 
            'utf8', 
            () => {}
        );
        console.log(localeFile, 'file saved.');
    });
});